gnome-image-installer/util/eos-reformatter.c

```
@@ -0,0 +1,972 @@
+    {
+      buf->len = 0;
+      return buf;
+    }
+
+  /* Initialize our pool, if it's not created yet */
+  if (reformatter->pool == NULL)
+    {
+      reformatter->pool = g_new0(guchar, reformatter->total_buffers * (reformatter->buffer_size + reformatter->page_size));
+    }
+
+  /* A bit awkwardly, exhaust doesn't apply when we create new buffers */
+  if (reformatter->used_buffers < reformatter->total_buffers)
+    {
+      /* No free buffers, but we can create one */
+      gsize offset = reformatter->used_buffers * (reformatter->buffer_size + reformatter->page_size);
+      guchar *unaligned = reformatter->pool + offset;
+      gsize page_size = reformatter->page_size;
+
+      buf = g_new0(EosAlignedBuffer, 1);
+      buf->ptr = (guchar*) (((gintptr) (unaligned + page_size)) & (~(page_size - 1)));
+      buf->len = 0;
+      reformatter->used_buffers++;
+    }
+  else
+    {
+      /* If we can't create a new buffer, we need to wait for one */
+      EosAlignedBuffer *buf;
+
+      while (!exhaust && g_async_queue_length (reformatter->free_queue) < half_buffers);
+        {
+          g_thread_yield();
+        }
+
+      do
+        {
+          g_thread_yield();
+          buf = g_async_queue_try_pop (reformatter->free_queue);
+        }
+      while (buf == NULL && reformatter->error == NULL);
+      buf->len = 0;
+      return buf;
+    }
+
+  return buf;
+}
+
+static gint
+eos_reformatter_prepare_read (EosReformatter *reformatter)
+{
+  GFileInfo *info;
+  GFile *file = g_file_new_for_path (reformatter->image);
+  GError *error = NULL;
+
+  info = g_file_query_info (file,
+                            G_FILE_ATTRIBUTE_STANDARD_CONTENT_TYPE ","
+                            G_FILE_ATTRIBUTE_STANDARD_SIZE,
+                            G_FILE_QUERY_INFO_NONE,
+                            NULL,
+                            &error);
+
+  if (info == NULL)
+    {
+      reformatter->error = g_error_new (EOS_REFORMATTER_ERROR, 0, _("Internal error"));
```

Wouldn't it be better to propagate the error from `g_file_query_info`, possibly prefixed with some extra information about the context, and have the error page examine the error domain to work out what to tell the user (with "Internal error" being the fallback for unexpected errors)?

---

I'm also not that wild about `reformatter->error` being set from potentially many threads without any locking.

---

Yes, propagating meaningful errors is definitely the intent, the error handling was a bit rushed in to move forward. Hence the fuzzy ownership as well. We'll probably want the error to be delivered either with the finished signal or separate error signal.